File: 15. 3Sum/solution.py
#    Runtime: 480 ms, faster than 96.80% of Python online submissions for 3Sum.
#    Memory Usage: 15.4 MB, less than 20.46% of Python online submissions for 3Sum.

class Solution(object):
    def threeSum(self, nums):
        nums.sort()
        res = []
        i = 0
        while i < len(nums):
            target = -nums[i]
            if target < 0:
                break

            begin = i + 1
            end = len(nums) - 1
            while begin < end:
                if nums[begin] + nums[end] > target:
                    end -= 1
                elif nums[begin] + nums[end] < target:
                    begin += 1
                elif nums[begin] + nums[end] == target:
                    res.append([-target, nums[begin], nums[end]])
                    while(begin < end and nums[end] == res[len(res)-1][2]):
                        end -= 1
                    while (begin < end and nums[begin] == res[len(res)-1][1]):
                        begin += 1
            while i + 1 < len(nums) and nums[i] == nums[i+1]:
                i += 1
            i += 1
        return res

File: 15. 3Sum/solution.cpp
/*
    Это не совсем моё решение. Моё было медленнее и не столь эффективно. Мне понравилось это решение, я его оптимизировал
    и добился отличной скорости.

    Runtime: 72 ms, faster than 100.00% of C++ online submissions for 3Sum.
    Memory Usage: 16.1 MB, less than 61.68% of C++ online submissions for 3Sum.
*/

class Solution {
public:
    vector<vector<int>> threeSum(vector<int> &num) {
    vector<vector<int>> res;
    std::sort(num.begin(), num.end());
    for (int i = 0; i < num.size(); i++) {
        int target = -num[i];
        int front = i + 1;
        int back = num.size() - 1;

        if(target < 0)
            break;

        while (front < back) {
            int sum = num[front] + num[back];
            if (sum < target)
                front++;
            else if (sum > target)
                back--;
            else {
                // конструируем вектор на месте
                res.push_back({num[i], num[front], num[back]});
                
                // убираем дубликаты:
                while (front < back && num[front] == res[res.size()-1][1]) 
                	front++;
                while (front < back && num[back] == res[res.size()-1][2]) 
                	back--;
            }
        }
        // убираем дубликаты:
        while (i + 1 < num.size() && num[i + 1] == num[i])
            i++;
    }
    return res;
}

};
